File: sdk-ts/src/circuitBreaker.ts
import { AgentFlowError } from "./exceptions.js";

export enum CircuitState {
  CLOSED = "closed",
  OPEN = "open",
  HALF_OPEN = "half_open",
}

export class CircuitOpenError extends AgentFlowError {
  constructor(message = "circuit is open") {
    super(message);
    this.name = "CircuitOpenError";
  }
}

export interface CircuitBreakerOptions {
  failureThreshold?: number;
  resetTimeoutMs?: number;
  halfOpenMaxCalls?: number;
}

export class CircuitBreaker {
  readonly failureThreshold: number;
  readonly resetTimeoutMs: number;
  readonly halfOpenMaxCalls: number;
  private _state = CircuitState.CLOSED;
  private failureCount = 0;
  private openedAt = 0;
  private halfOpenCalls = 0;

  constructor(options: CircuitBreakerOptions = {}) {
    this.failureThreshold = options.failureThreshold ?? 5;
    this.resetTimeoutMs = options.resetTimeoutMs ?? 30_000;
    this.halfOpenMaxCalls = options.halfOpenMaxCalls ?? 1;
  }

  beforeCall(): void {
    if (this._state === CircuitState.OPEN) {
      if (Date.now() - this.openedAt >= this.resetTimeoutMs) {
        this._state = CircuitState.HALF_OPEN;
        this.halfOpenCalls = 0;
      } else {
        throw new CircuitOpenError("circuit is open");
      }
    }
    if (this._state === CircuitState.HALF_OPEN) {
      if (this.halfOpenCalls >= this.halfOpenMaxCalls) {
        throw new CircuitOpenError("circuit is half-open, probe in flight");
      }
      this.halfOpenCalls += 1;
    }
  }

  recordSuccess(): void {
    this._state = CircuitState.CLOSED;
    this.failureCount = 0;
    this.halfOpenCalls = 0;
  }

  recordFailure(): void {
    if (this._state === CircuitState.HALF_OPEN) {
      this._state = CircuitState.OPEN;
      this.openedAt = Date.now();
      this.halfOpenCalls = 0;
      return;
    }
    this.failureCount += 1;
    if (this.failureCount >= this.failureThreshold) {
      this._state = CircuitState.OPEN;
      this.openedAt = Date.now();
      this.halfOpenCalls = 0;
    }
  }

  get state(): CircuitState {
    return this._state;
  }
}

File: sdk/agentflow/circuit_breaker.py
import time
from dataclasses import dataclass, field
from enum import Enum
from threading import Lock

from agentflow.exceptions import AgentFlowError


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class CircuitOpenError(AgentFlowError):
    pass


@dataclass
class CircuitBreaker:
    failure_threshold: int = 5
    reset_timeout_s: float = 30.0
    half_open_max_calls: int = 1

    _state: CircuitState = field(default=CircuitState.CLOSED, init=False)
    _failure_count: int = field(default=0, init=False)
    _opened_at: float = field(default=0.0, init=False)
    _half_open_calls: int = field(default=0, init=False)
    _lock: Lock = field(default_factory=Lock, init=False)

    def before_call(self) -> None:
        with self._lock:
            if self._state == CircuitState.OPEN:
                if time.monotonic() - self._opened_at >= self.reset_timeout_s:
                    self._state = CircuitState.HALF_OPEN
                    self._half_open_calls = 0
                else:
                    raise CircuitOpenError("circuit is open")
            if self._state == CircuitState.HALF_OPEN:
                if self._half_open_calls >= self.half_open_max_calls:
                    raise CircuitOpenError("circuit is half-open, probe in flight")
                self._half_open_calls += 1

    def record_success(self) -> None:
        with self._lock:
            self._state = CircuitState.CLOSED
            self._failure_count = 0
            self._half_open_calls = 0

    def record_failure(self) -> None:
        with self._lock:
            if self._state == CircuitState.HALF_OPEN:
                self._state = CircuitState.OPEN
                self._opened_at = time.monotonic()
                self._half_open_calls = 0
                return
            self._failure_count += 1
            if self._failure_count >= self.failure_threshold:
                self._state = CircuitState.OPEN
                self._opened_at = time.monotonic()
                self._half_open_calls = 0

    @property
    def state(self) -> CircuitState:
        return self._state
